rmw_fastrtps_dynamic_cpp/include/rmw_fastrtps_dynamic_cpp/TypeSupport_impl.hpp

```diff
@@ -437,26 +289,31 @@ bool TypeSupport<MembersType>::serializeROSmessage(
           if (!member->is_array_) {
             serializeROSmessage(ser, sub_members, field);
           } else {
-            void * subros_message = nullptr;
             size_t array_size = 0;
-            size_t sub_members_size = sub_members->size_of_;
-            size_t max_align = calculateMaxAlign(sub_members);
 
             if (member->array_size_ && !member->is_upper_bound_) {
-              subros_message = field;
               array_size = member->array_size_;
             } else {
-              array_size = get_array_size_and_assign_field(
-                member, field, subros_message, sub_members_size, max_align);
+              if (!member->size_function) {
+                RMW_SET_ERROR_MSG("unexpected error: size function is null");
+                return false;
+              }
+              array_size = member->size_function(field);
```

Comment:
does this work?
I think that the `size_function` is for querying the current size, here you want the desired size.
See https://github.com/ros2/rmw_cyclonedds/blob/91aed377a0a2e39664d0eae401ee49e9c419540a/rmw_cyclonedds_cpp/include/rmw_cyclonedds_cpp/TypeSupport_impl.hpp#L375.

Reply:
I'm sorry, I'm reading this incorrectly.
This is the serializing function, not the deserializing one.

Out of curiosity, why is this needed here and not in `rmw_cyclonedds`?

Reply:
`rmw_cyclonedds` does have a custom serializer, but it is on another source file ([this one](https://github.com/ros2/rmw_cyclonedds/blob/91aed377a0a2e39664d0eae401ee49e9c419540a/rmw_cyclonedds_cpp/src/serdata.cpp))